docs/architecture/pk-4.md

```diff
@@ -0,0 +1,352 @@
+
+* Arguably, since PK is now specialized towards signatures, we could remove the `usage` argument from `mbedtls_pk_get_psa_attributes()`, and make the function work only for a sign/verify usage. However, I don't think this would be right, especially because PK only handles signature: after parsing a key, to use it for something other than signature, you need to use PK, which means you need to call `mbedtls_pk_get_psa_attributes()` then `mbedtls_pk_import_into_psa()` to create a PSA key with your desired non-signature algorithm. Also, the usage parameter allows for extracting the public part of a key when you don't know whether you have a public key or a key pair, which is very convenient in some cases such as managing a key store.
+
+#### Copy between PK and PSA
+
+Keep the following functions, which create a PK context from a PSA key and vice versa:
+```
+mbedtls_pk_import_into_psa()
+mbedtls_pk_copy_from_psa()
+mbedtls_pk_copy_public_from_psa()
+```
+
+#### Wrap a PSA key in PK
+
+There is already a function to wrap a PSA private key in a PK context: `mbedtls_pk_setup_opaque()`. The function's name no longer makes sense, since there is no longer a concept of “opaque” PK contexts, and no longer a ”setup“ operation on PK contexts.
+
+ACTION: rename `mbedtls_pk_setup_opaque()` to `mbedtls_pk_wrap_psa()` and adjust the documentation accordingly. (Updating internal references to “opaque” is out of scope and will be done later: [Update terminology from “opaque” to “wrapped”](#update-terminology-from-opaque-to-wrapped).)
+
+ACTION: remove mentions of operations other than sign and verify from the documentation.
+
+The current function has some limitations. We should lift them soon (“[Remove limitations of `mbedtls_pk_wrap_psa()`](#remove-limitations-of-mbedtls_pk_wrap_psa)”), but it isn't a deal breaker for TF-PSA-Crypto 1.0.
+
+### Key parsing and writing
+
+Keep:
+```
+mbedtls_pk_parse_key()
+mbedtls_pk_parse_public_key()
+mbedtls_pk_write_key_der()
+mbedtls_pk_write_pubkey_der()
+mbedtls_pk_write_pubkey_pem()
+mbedtls_pk_write_key_pem()
+```
+
+Keep:
+```
+mbedtls_pk_parse_keyfile()
+mbedtls_pk_parse_public_keyfile()
+```
+
+### Signature conveniences
+
+#### Basic signature functionality
+
+Keep the following:
+```
+MBEDTLS_PK_SIGNATURE_MAX_SIZE
+mbedtls_pk_verify()
+mbedtls_pk_sign()
+```
+
+ACTION: remove `MBEDTLS_ERR_PK_SIG_LEN_MISMATCH`.
+
+#### Restartable signature functionality
+
+It's more convenient to keep using PK for restartable signature in X.509, for the same reason as non-restartable signature. So restartable PK should keep existing, so we might as well keep it public.
```

Comment:
Questions:
- guarded by what? (Currently `ECDSA_C && `ECP_RESTARTABLE`.)
- these functions are currently implemented on top of the ECP API, how confident are we that they can be efficiently implemented on top of the PSA API?

Reply:
Keep the current guards. We might tweak them, but they are currently “there is a built-in implementation of restartable ECC”, and I don't foresee any evolution other than generalizing to non-ECC or non-built-in.

PK is deeply embedded in X.509, so restartable PK had better be a reasonable way to do restartable verification. I haven't thought about this a lot. Do you think there's a risk here, that we'd want to change the function prototypes? Should we make these functions private until we've changed their implementation to go via PSA?

(There's of course the question of whether we want to ever add a PSA layer, but I think we do, because there is interest in the restartable API with hardware-specific implementations.)